keps/sig-network/2004-topology-aware-routing/README.md

```diff
@@ -0,0 +1,600 @@
+   most insufficient endpoints first.
+3. Iterate through the `endpoints_available` pool and the `endpoints_require`
+   pool. Pop out the first zones from both pools and assign one endpoint from
+   the `available_zone` to the `require_zone`. Push the `available_zone` back
+   to the `endpoints_available` pool if it still has overload below threshold.
+   Push the `require_zone` back to the `endpoints_require pool` if it still has
+   overload above threshold. This step ends when one of the conditions meets:
+   - `endpoints_available` pool is empty.
+   - `endpoints_require` pool is empty.
+4. If `endpoints_available` pool is empty and `endpoints_require` pool is not
+   empty, downgrade to `Balanced` approach.
+5. Otherwise, iterate all zones again, assign endpoints from zones in
+   `endpoints_available` pool to zones with endpoints fewer than expected
+   endpoints. In this step, zones in `endpoints_available` pool only provide
+   extra endpoints when they have endpoints > expected endpoints. This ends
+   when one of the conditions meets:
+   - `endpoints_available` pool is empty.
+   - All zones have at least the flooring number of expected endpoints in their
+  local SliceGroup.
+
+ZoneA, ZoneB, ZoneC expect 3.2, 1.6, 0.2 endpoints respectively.
+
+<img
+src="https://github.com/robscott/k8s-enhancements/blob/topology-aware-routing/keps/sig-network/2004-topology-aware-routing/ep-distribution.png"
+alt="ep distribution" width="600">
+
+ZoneA gives one extra endpoint to ZoneC to help ZoneC maintain overload.
+
+<img
+src="https://github.com/robscott/k8s-enhancements/blob/topology-aware-routing/keps/sig-network/2004-topology-aware-routing/ep-distribution2.png"
+alt="ep distribution" width="600">
+
+ZoneA has to give one more endpoint to ZoneB to help ZoneB maintain overload.
+
+<img
+src="https://github.com/robscott/k8s-enhancements/blob/topology-aware-routing/keps/sig-network/2004-topology-aware-routing/ep-distribution3.png"
+alt="ep distribution" width="600">
+
+ZoneA ends with endpoints fewer than expected. But its overload is still below
+threshold.
+
+<img
+src="https://github.com/robscott/k8s-enhancements/blob/topology-aware-routing/keps/sig-network/2004-topology-aware-routing/ep-distribution4.png"
+alt="ep distribution" width="600">
+
+During our tests among different datasets, we found that with a large number of
+total endpoints it is less likely to have cases described in step 4, that's one
+of the reasons we introduced the starting threshold for this mechanism.
+Meanwhile as described above, we may open the starting threshold as a
+configurable parameter for users. This might be updated after collecting enough
+feedback from users.
+
+#### Rationale
+1. Easy to implement and acceptable churn of EndpointSlices
+2. Relatively high in-zone traffic
+3. Balanced deviation for endpoints
+4. No weights involved -- no API additions needed
+
+#### Potential Issues
+1. Can't deal with the single-zone traffic case described above
```

Comment:
but you work around it by degrading, right?

Reply:
This was confusing - we were trying to describe the possibility of overloaded endpoints if a large portion of traffic originates from a single zone. I've updated this to be more clear. You're right that we would downgrade if a single-zone simply had far too many nodes/endpoints to effectively balance - we just don't have any feedback available to downgrade based on where traffic ends up originating from.